docs/design/K23SI_TXN.md

```diff
@@ -0,0 +1,516 @@
+- All write operations specify a TRH in addition to the MTR.
+- The participant who receives a write creates a WI and a TWIM record which tracks all of the WIs on that participant. The TWIM record contains the TRH key so if another txn creates a conflict, we can consult with the TRH to determine a transaction to abort.
+- The TWIM record can also produce a list of the WIs created on behalf of a transaction locally so that it can convert WIs when the TRH comes to finalize a transaction (i.e. after the application commits/aborts).
+- The TWIMs also help cleanup txns should they become abandoned.
+
+### Reads
+![Read](../images/TxnRead.png)
+Read operations communicate a `key` which is expressed as an SKVRecord. That is, the fields in the SKVRecord which comprise the partitioning and range keys must have specified values. The client has to find out where to send the request, based on the current partition map (not shown here since it isn't relevant to the txn algorithm). The request is then sent to the owner node of the data, who records the data access in its read cache and responds to client with success/failure.
+The read operations are annotated with an MTR. The timestamp in this MTR is used as the snapshot version(SV) for MVCC snapshot read purposes. The reads are standard MVCC reads i.e. the returned value is the latest value such that `result.ts <= request.MTR.Timestamp`.
+
+##### Read Cache
+In order to prevent various anomalies and achieve SerializableSI, we maintain a read key cache. The purpose of the cache is to remember the latest observed version for every key in the system so that no-one will be allowed to modify the observed history of the system. This cache is maintained in memory at each node/partition.
+- When a read is received for a transaction `T`, we remember that the given key was observed at timestamp `T.timestamp`. If we've previously observed a read for the same key, we update the cache to reflect the max such value and the latest transaction to observe it.
+- The implementation of the read key cache is an interval tree. We require an interval tree to also remember key range reads.
+- The cache is consulted at write time to determine if a write should be aborted. The reason we need to do this is that if we try to write an item with timestamp <= lastSVTimeTheKeyWasRead, then we are breaking a promise to whoever read the item - they saw some item version when they read at their snapshot time, and now we're trying to insert a newer version into their snapshot. This write should therefore be aborted.
+- Entries are removed from the cache in an LRU fashion. We maintain a minSVTimestamp watermark for the entire cache, which tells us how old is the oldest entry in the cache. Any write before this timestamp (for any key) is aborted as we assume there may have been a read for it.
```

Comment:
Not a big deal, but it is technically not an LRU. We always evict the oldest entry by insertion order, cache hits don't change the eviction order.

Reply:
true. Reworded.